x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java

```diff
@@ -176,14 +176,20 @@ public void killProcess(JobTask jobTask, boolean awaitCompletion, String reason)
             awaitCompletion,
             reason
         ));
+        jobTask.closing();
```

Comment:
The thing that worries me about this line is that I think we use kill as part of the graceful node shutdown sequence, where we want to leave the job open despite killing the process so that it starts up on another node after the current node leaves the cluster.

I haven’t looked closely because I’m on mobile at the moment, so I may be wrong and maybe the graceful shutdown sequence goes through a different code path. I will have a more thorough look tomorrow.

Reply:
for sure

Reply:
So, this `closing` value is not persisted. It is only useful when a task is already constructed on a given node.

This will be OK in the graceful shutdown as it will:

 - Allow a job to be killed while it is starting up (it will not set the state to closed or blow away the persistent task)
 - Locking the context seems like a necessity as otherwise, the `kill` will fail in the window of the context NOT having fully constructed the processing message handler and the job starting up.

Maybe the place to fix this is in the `close` logic. I will see if there is a way to fix the path there instead (so that close waits, or sets the closing flag in a better way)

Reply:
The issue on `close` is that we do indeed set `isClosing`, but since there is no process context (kill deleted it from the map), close effectively takes no action.

But, a reference to the process context still exists (outside the map), within the start up logic 🤦

Reply:
OK, `kill` needs to be able to run while the job is closing. 

So, locking the process context doesn't seem like an option. But, we need to somehow prevent the process context reference that exists during open from starting if it was killed (even if the communicator doesn't yet exist).